src/safari-client.js

```diff
@@ -0,0 +1,222 @@
+        'Pusher Beams does not support this safari version (Safari Push Noifications not supported)'
+      );
+    }
+
+    this.instanceId = instanceId;
+    this._deviceId = null;
+    this._token = null;
+    this._userId = null;
+    this._deviceStateStore = new DeviceStateStore(instanceId);
+    this._endpoint = endpointOverride; // Internal only
+    this._platform = 'safari';
+
+    this._ready = this._init();
+  }
+
+  async _init() {
+    if (this._deviceId !== null) {
+      return;
+    }
+
+    await this._deviceStateStore.connect();
+
+    await this._detectSubscriptionChange();
+
+    this._deviceId = await this._deviceStateStore.getDeviceId();
+    this._token = await this._deviceStateStore.getToken();
+    this._userId = await this._deviceStateStore.getUserId();
+  }
+
+  async _detectSubscriptionChange() {
+    const storedToken = await this._deviceStateStore.getToken();
+    const actualToken = getDeviceToken();
+
+    const tokenHasChanged = storedToken !== actualToken;
+    if (tokenHasChanged) {
+      // The device token has changed. This is should only really happen when
+      // users restore from an iCloud backup
+      await this._deviceStateStore.clear();
+      this._deviceId = null;
+      this._token = null;
+      this._userId = null;
+    }
+  }
+
+  _requestPermission() {
+    return new Promise((resolve) => {
+      window.safari.pushNotification.requestPermission(
+        __url,
+        __pushId,
+        { userID: 'abcdef' },
+        resolve
+      );
+    });
+  }
+
+  async start() {
+    if (this._deviceId !== null) {
+      return this;
+    }
+
+    let { deviceToken, permission } = getPermission(__pushId);
+
+    if (permission === 'default') {
+      console.debug('permission is default, requesting permission');
+      let { deviceToken, permission } = await this._requestPermission(__pushId);
+      if (permission == 'granted') {
+        const deviceId = await this._registerDevice(deviceToken);
+        await this._deviceStateStore.setToken(deviceToken);
+        await this._deviceStateStore.setDeviceId(deviceId);
+        await this._deviceStateStore.setLastSeenSdkVersion(sdkVersion);
+        await this._deviceStateStore.setLastSeenUserAgent(
+          window.navigator.userAgent
+        );
+        this._token = deviceToken;
+        this._deviceId = deviceId;
+      }
+    }
+  }
+
+  async getRegistrationState() {
+    await this._resolveSDKState();
+
+    const { permission } = getPermission(__pushId);
+
+    if (permission === 'denied') {
+      return RegistrationState.PERMISSION_DENIED;
+    }
+
+    if (permission === 'granted' && this._deviceId !== null) {
+      return RegistrationState.PERMISSION_GRANTED_REGISTERED_WITH_BEAMS;
+    }
+
+    if (permission === 'granted' && this._deviceId === null) {
+      return RegistrationState.PERMISSION_GRANTED_NOT_REGISTERED_WITH_BEAMS;
+    }
+
+    return RegistrationState.PERMISSION_PROMPT_REQUIRED;
+  }
+
+  async clearAllState() {
```

Comment:
I'm not sure what the purpose of this method is for non-safari web notifications 🤔  I doubt you would want to show the permission request dialog box when clearing the state. I guess the behaviour of this method only makes sense if it assumes that the user has already given permission? And therefore it probably doesn't matter that this won't successfully ask for permission in Safari, because developers probably wouldn't want to do that. If that is the assumption then we should probably make that explicit with checks/errors. I can explain that if it didn't make sense.